diff -ur tbb-2.0r014/build/Makefile.tbb tbb-2.0r014.test/build/Makefile.tbb
--- tbb-2.0r014/build/Makefile.tbb	2007-09-21 21:16:34.000000000 -0400
+++ tbb-2.0r014.test/build/Makefile.tbb	2007-12-15 17:28:09.000000000 -0500
@@ -74,9 +74,12 @@
 
 # Object files that gmake up TBB (TBB_ASM.OBJ is platform-specific)
 TBB.OBJ = $(TBB_CPLUS.OBJ) $(TBB_ASM.OBJ)
+LIB_LINK_FLAGS += -Wl,-soname=$(TBB.DLL).$(SONAME_SUFFIX)
 
 $(TBB.DLL): $(TBB.OBJ) $(TBB.DEF) $(TBB.RES) tbbvars
-	$(CPLUS) $(TBB.OBJ) $(TBB.RES) $(LIBS) $(PIC_KEY) $(OUTPUT_KEY)$(TBB.DLL) $(LIB_LINK_FLAGS)
+	$(CPLUS) $(TBB.OBJ) $(TBB.RES) $(LIBS) $(PIC_KEY) $(OUTPUT_KEY)$(TBB.DLL).$(VERSION) $(LIB_LINK_FLAGS)
+	ln -s $(TBB.DLL).$(VERSION) $(TBB.DLL).$(SONAME_SUFFIX)
+	ln -s $(TBB.DLL).$(VERSION) $(TBB.DLL)
 
 
 #clean:
diff -ur tbb-2.0r014/build/Makefile.tbbmalloc tbb-2.0r014.test/build/Makefile.tbbmalloc
--- tbb-2.0r014/build/Makefile.tbbmalloc	2007-09-21 21:16:34.000000000 -0400
+++ tbb-2.0r014.test/build/Makefile.tbbmalloc	2007-12-15 17:28:32.000000000 -0500
@@ -58,6 +58,7 @@
 # MALLOC_CPLUS.OBJ is built in two steps due to Intel Compiler Tracker # C69574
 MALLOC.OBJ := $(MALLOC_CPLUS.OBJ) $(MALLOC_ASM.OBJ) $(MALLOC_CUSTOM.OBJ) MemoryAllocator.$(OBJ)
 MALLOC_CPLUS.OBJ += MemoryAllocator.$(OBJ)
+LIB_LINK_FLAGS += -Wl,-soname=$(MALLOC.DLL).$(SONAME_SUFFIX)
 M_CPLUS_FLAGS := $(subst $(WARNING_KEY),,$(M_CPLUS_FLAGS))
 
 $(MALLOC_CPLUS.OBJ): %.$(OBJ): %.cpp
@@ -69,7 +70,9 @@
 
 $(MALLOC.DLL): TBB.DEF=$(MALLOC.DEF)
 $(MALLOC.DLL): $(MALLOC.OBJ) $(MALLOC.DEF) $(TBBMALLOC.RES)
-	$(CPLUS) $(MALLOC.OBJ) $(TBBMALLOC.RES) $(LIBS) $(PIC_KEY) $(OUTPUT_KEY)$(MALLOC.DLL) $(LIB_LINK_FLAGS)
+	$(CPLUS) $(MALLOC.OBJ) $(TBBMALLOC.RES) $(LIBS) $(PIC_KEY) $(OUTPUT_KEY)$(MALLOC.DLL).$(VERSION) $(LIB_LINK_FLAGS)
+	ln -s $(MALLOC.DLL).$(VERSION) $(MALLOC.DLL).$(SONAME_SUFFIX)
+	ln -s $(MALLOC.DLL).$(VERSION) $(MALLOC.DLL)
 
 malloc: $(MALLOC.DLL)
 
diff -ur tbb-2.0r014/build/common.inc tbb-2.0r014.test/build/common.inc
--- tbb-2.0r014/build/common.inc	2007-09-21 21:16:34.000000000 -0400
+++ tbb-2.0r014.test/build/common.inc	2007-12-15 16:43:10.000000000 -0500
@@ -24,6 +24,9 @@
 # invalidate any other reasons why the executable file might be covered by
 # the GNU General Public License.
 
+VERSION=2.0.014
+SONAME_SUFFIX=2
+
 ifndef tbb_os
  ifeq ($(OS), Windows_NT)
   export tbb_os=windows
