diff -urp tbb20_20070927oss_src/Makefile tbb20_20070927oss_src.pm2/Makefile
--- tbb20_20070927oss_src/Makefile	2007-09-28 22:13:21.000000000 +0200
+++ tbb20_20070927oss_src.pm2/Makefile	2007-12-19 19:46:26.000000000 +0100
@@ -25,6 +25,7 @@
 # the GNU General Public License.
 
 tbb_root?=.
+orig_tbb_root:=$(tbb_root)
 include $(tbb_root)/build/common.inc
 .PHONY: all tbb tbbmalloc test debug examples clean
 
@@ -47,6 +48,9 @@ examples: tbb tbbmalloc examples_debug c
 clean: clean_release clean_debug clean_examples
 	@echo clean done
 
+# Introduce explicit dependencies to aid parallel make.
+test_release: tbb_release
+tbbmalloc_test_release: tbbmalloc_release
 
 .PHONY: tbb_release tbb_debug test_release test_debug
 
@@ -57,11 +60,15 @@ tbb_release: mkdir_release
 tbb_debug: mkdir_debug
 	$(MAKE) -C "$(work_dir)_debug"  -r -f $(tbb_root)/build/Makefile.tbb cfg=debug tbb_root=$(tbb_root)
 
-test_release: mkdir_release tbb_release test_release_no_depends
+test_release: mkdir_release tbb_release
+	$(MAKE) cfg=release tbb_root=$(orig_tbb_root) test_release_no_depends
+
 test_release_no_depends: 
 	-$(MAKE) -C "$(work_dir)_release"  -r -f $(tbb_root)/build/Makefile.test cfg=release tbb_root=$(tbb_root) 
 
-test_debug: tbb_debug mkdir_debug test_debug_no_depends
+test_debug: tbb_debug mkdir_debug
+	$(MAKE) cfg=release tbb_root=$(orig_tbb_root) test_debug_no_depends
+
 test_debug_no_depends:
 	-$(MAKE) -C "$(work_dir)_debug"  -r -f $(tbb_root)/build/Makefile.test cfg=debug tbb_root=$(tbb_root)
 
@@ -73,11 +80,15 @@ tbbmalloc_release: mkdir_release
 tbbmalloc_debug: mkdir_debug
 	$(MAKE) -C "$(work_dir)_debug"  -r -f $(tbb_root)/build/Makefile.tbbmalloc cfg=debug malloc tbb_root=$(tbb_root)
 
-tbbmalloc_test_release: tbb_release tbbmalloc_release mkdir_release tbbmalloc_test_release_no_depends
+tbbmalloc_test_release: tbb_release tbbmalloc_release mkdir_release
+	$(MAKE) cfg=release tbb_root=$(orig_tbb_root) tbbmalloc_test_release_no_depends
+
 tbbmalloc_test_release_no_depends:
 	-$(MAKE) -C "$(work_dir)_release"  -r -f $(tbb_root)/build/Makefile.tbbmalloc cfg=release malloc_test tbb_root=$(tbb_root)
 
-tbbmalloc_test_debug: tbb_debug tbbmalloc_debug mkdir_debug tbbmalloc_test_debug_no_depends
+tbbmalloc_test_debug: tbb_debug tbbmalloc_debug mkdir_debug
+	$(MAKE) cfg=release tbb_root=$(orig_tbb_root) tbbmalloc_test_debug_no_depends
+
 tbbmalloc_test_debug_no_depends:
 	-$(MAKE) -C "$(work_dir)_debug"  -r -f $(tbb_root)/build/Makefile.tbbmalloc cfg=debug malloc_test tbb_root=$(tbb_root)
 
